Pwr 259

ntenelsen opened the pull request:
Fixed an issue to now succeed in both of the tests responsible for failing the tests in PWR-272. One test is "successful", but always results in a success due to it's design, the other test was changed to be fulfillable with the outlined requirements.

Therefore it's not entirely clear what the requirements for updateProfileSkills in the profile service are. Previously one test (which seems to require rather nonsensical logic anyways) would always succeed while the other would fail even when the success condition outlined in it's function name were met.

ntenelsen committed:
[pwr-259] changed skill updating to always rely on the client properly verifying input first.

Not properly tested yet.

ntenelsen committed:
[PWR-259] fixed updateProfileSkills to always remove old variants of the same skill and overwrite with the new version

Changed shouldUpdateSkillRating_whenNewRatingHigher() to actually tests what it's name implies.

ShouldNotUpdateSkillRating_whenNewRatingLower() is still broken.

Niklas345 commented on src/main/java/de/hbt/pwr/profile/controller/ProfileEntryEndpoint.java:
```diff
@@ -195,6 +195,7 @@ public void deleteEducationEntry(@PathVariable("initials") String initials, @Pat
 
     @PutMapping("/skill")
     public Skill updateSkill(@PathVariable("initials") String initials, @RequestBody Skill skill) {
+        System.out.println("Skill update request");
```
Bitte auf log.debug umstellen

Niklas345 commented on src/main/java/de/hbt/pwr/profile/service/ProfileEntryService.java:
```diff
@@ -69,12 +69,11 @@ public void deleteEntryWithId(Long id, Profile profile, NameEntityType nameEntit
+        Set<Skill> concurrent = profile.getSkills()
+                .stream().filter(s -> hasEqualName(s, skill))
+                .collect(Collectors.toSet());
+        concurrent.stream().forEach(s -> profile.getSkills().remove(s));
```
den .stream() call kannst du dir hier sparen. Auf einer collection kannst du immer .forEach() aufrufen

Niklas345 commented on src/main/java/de/hbt/pwr/profile/service/ProfileEntryService.java:
```diff
@@ -69,12 +69,11 @@ public void deleteEntryWithId(Long id, Profile profile, NameEntityType nameEntit
+                .collect(Collectors.toSet());
+        concurrent.stream().forEach(s -> profile.getSkills().remove(s));
+        return createNewInProfile(skill, profile);
     }
```
Optional<Skill> concurrent = profile.getSkills()
                .stream().filter(s -> hasEqualName(s, skill))
                .findAny();
        return concurrent
                .map(s -> updateSkill(s, skill))
                .orElseGet(() -> createNewInProfile(skill, profile));

Mit dieser Logik klappt es wieder. Ich verstehe den fix, aber er ist unnötig. Du hast in einem vorherigen commit die Logik ausgebaut gehabt, dass Dinge gemerged werden.
Es sollte nicht notwendig sein, hier mehrere concurrent skills zu finden, da es nur einen geben kann. Alles andere sollte rausgefiltert werden (schon vorher).
Was das hier macht:

- Duplikate Finden
- Alle Duplikat entfernen
- Skill neu anlegen

Das dauernde Skill neu anlegen sollte überflüssig sein.

ntenelsen committed:
[PWR-259] Implemented change requests.
Changed updateProfileSkills logic to be more concise.
Changed updateSkill so it also updates versions.
